### PR TITLE
feat: restore support for SSJ

### PR DESCRIPTION
**Pull request by enitrat**
# Pull Request type

Time spent on this PR:

Resolves: #702

Please check the type of change your PR introduces:

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Testing

## Does this introduce a breaking change?

- [ ] Yes
- [ ] No

### CODE REVIEW

File: crates/ef-testing/src/evm_sequencer/account/mod.rs [hunk 1]
```diff
@@ -111,4 +149,23 @@ mod tests {
         // Then
         assert_eq!(result, vec![StarkFelt::from(0x0102030405u64)]);
     }
+
+    #[test]
+    fn test_inner_byte_array_pointer() {
+        // Given
+        let base_address: StarkFelt = get_storage_var_address(ACCOUNT_BYTECODE, &[]).into();
+        let chunk = FieldElement::from(0_u32);
```
Comment (tcoratger) [inline crates/ef-testing/src/evm_sequencer/account/mod.rs:157]:
```suggestion
        let chunk = FieldElement::ZERO;
```

File: crates/ef-testing/src/evm_sequencer/account/mod.rs [hunk 2]
```diff
@@ -111,4 +149,23 @@ mod tests {
...
+
+        // When
+        let result = inner_byte_array_pointer(FieldElement::from(base_address), chunk);
```
Comment (tcoratger) [inline crates/ef-testing/src/evm_sequencer/account/mod.rs:160]:
```suggestion
        let result = inner_byte_array_pointer(base_address.into(), chunk);
```

File: crates/ef-testing/src/evm_sequencer/account/v1.rs [hunk 3]
```diff
@@ -7,15 +7,57 @@ use starknet_api::{
...
+    let mut bytearray = vec![(bytecode_base_address, StarkFelt::from(code.len() as u128))];
+
+    let mut bytecode_storage = split_bytecode_to_starkfelt(&code)
+        .enumerate()
+        .map(|(index, b)| {
+            let offset = index % 256;
+            let index = index / 256;
+            let key = inner_byte_array_pointer(
+                FieldElement::from(*bytecode_base_address.0.key()),
+                FieldElement::from(index),
```
Comment (tcoratger) [inline crates/ef-testing/src/evm_sequencer/account/v1.rs:45]:
```suggestion
                (*bytecode_base_address.0.key()).into(),
                index.into(),
```

File: crates/ef-testing/src/evm_sequencer/account/v1.rs [hunk 4]
```diff
@@ -7,15 +7,57 @@ use starknet_api::{
...
+            );
+            (
+                offset_storage_key(
+                    StorageKey(PatriciaKey::try_from(StarkFelt::from(key)).unwrap()),
+                    offset as i64,
+                ),
+                b,
+            )
+        })
+        .collect::<Vec<_>>();
```
Comment (tcoratger) [inline crates/ef-testing/src/evm_sequencer/account/v1.rs:55]:
```suggestion
        .collect();
```

File: crates/ef-testing/src/evm_sequencer/account/v1.rs [hunk 5]
```diff
@@ -113,3 +119,38 @@ impl KakarotAccount {
...
+
+        // Then
+        let expected_result = vec![
+            (bytecode_base_address, StarkFelt::from(code.len() as u128)),
+            (
+                offset_storage_key(
+                    StorageKey(
+                        PatriciaKey::try_from(StarkFelt::from(inner_byte_array_pointer(
+                            FieldElement::from(*bytecode_base_address.0.key()),
+                            FieldElement::from(0_u32),
```
Comment (tcoratger) [inline crates/ef-testing/src/evm_sequencer/account/v1.rs:144]:
```suggestion
                           (*bytecode_base_address.0.key()).into(),
                            FieldElement::ZERO,
```

File: crates/ef-testing/src/evm_sequencer/evm_state/v1.rs [hunk 6]
```diff
@@ -186,52 +231,57 @@ impl Evm for KakarotSequencer {
...
-            .get_storage_at(starknet_address, pending_word_address)?;
-        bytecode.append(
-            &mut FieldElement::from(pending_word).to_bytes_be()[32 - pending_word_len..].to_vec(),
-        );
+        if pending_word_len != 0 {
+            let storage_chunk_index = num_chunks / 256;
+            let offset_in_chunk = num_chunks % 256;
+            let storage_pointer = inner_byte_array_pointer(
+                FieldElement::from(*bytecode_base_address.0.key()),
+                FieldElement::from(storage_chunk_index),
```
Comment (greged93) [inline crates/ef-testing/src/evm_sequencer/evm_state/v1.rs:271]: think @tcoratger for some `into()` here

File: crates/ef-testing/src/evm_sequencer/evm_state/v1.rs [hunk 7]
```diff
@@ -322,16 +372,20 @@ pub(crate) fn offset_storage_key(key: StorageKey, offset: i64) -> StorageKey {
...
+            },
+            sequencer::{KakarotEnvironment, INITIAL_SEQUENCER_STATE},
+            utils::compute_starknet_address,
         },
-        sequencer::{KakarotEnvironment, INITIAL_SEQUENCER_STATE},
+        models::result::extract_output_and_log_execution_result,
     };
     use reth_primitives::{
-        sign_message, AccessList, Signature, TransactionSigned, TxEip1559, TxValue, B256,
+        sign_message, AccessList, Signature, TransactionSigned, TxEip1559, TxLegacy, B256,
```
Comment (greged93) [inline crates/ef-testing/src/evm_sequencer/evm_state/v1.rs:388]: linter says `AccessList` and `TxEip1559` aren't used anymore

File: crates/ef-testing/src/evm_sequencer/evm_state/v1.rs [hunk 8]
```diff
@@ -420,55 +474,69 @@ mod tests {
...
                 nonce: 0,
                 gas_limit: 1_000_000,
-                max_fee_per_gas: 0,
-                max_priority_fee_per_gas: 0,
+                gas_price: 0,
                 to: reth_primitives::TransactionKind::Call(*TEST_CONTRACT_ADDRESS),
-                value: TxValue::from(U256::ZERO),
-                access_list: AccessList::default(),
+                value: U256::ZERO,
                 input: Bytes::default(),
```
Comment (greged93) [inline crates/ef-testing/src/evm_sequencer/evm_state/v1.rs:484]: you can remove `nonce`, `gas_price`, `value` and `input` and replace them with:
```rust
reth_primitives::Transaction::Legacy(TxLegacy {
                chain_id: Some(CHAIN_ID),
                gas_limit: 1_000_000,
                to: reth_primitives::TransactionKind::Call(*TEST_CONTRACT_ADDRESS),
                .. Default::default()
            }),
```

File: crates/ef-testing/src/evm_sequencer/evm_state/v1.rs [hunk 9]
```diff
@@ -420,55 +474,69 @@ mod tests {
...
+        let storage = sequencer
+            .state_mut()
+            .get_storage_at(
+                contract_starknet_address,
+                get_storage_var_address(
+                    ACCOUNT_STORAGE,
+                    &[StarkFelt::from(0u8), StarkFelt::from(0u8)],
+                ),
+            )
+            .unwrap();
```
Comment (greged93) [inline crates/ef-testing/src/evm_sequencer/evm_state/v1.rs:539]: this can just be `sequencer.storage_at(&TEST_CONTRACT_ADDRESS, U256::ZERO)`

**Review by greged93 — COMMENTED** [review]
couple of nits

File: crates/ef-testing/src/evm_sequencer/evm_state/v1.rs [hunk 10]
```diff
@@ -162,12 +202,17 @@ impl Evm for KakarotSequencer {
...
 
-        let nonce = if class_hash == self.environment.eoa_class_hash {
-            self.state_mut().get_nonce_at(starknet_address)?.0
-        } else if class_hash == self.environment.contract_account_class_hash {
-            let key = get_storage_var_address("contract_account_nonce", &[]);
+        //TODO: remove CA - EOA distinction
+        let nonce = if implementation == self.environment.contract_account_class_hash.0 {
+            let key = get_storage_var_address(ACCOUNT_NONCE, &[]);
             self.state_mut().get_storage_at(starknet_address, key)?
         } else {
```
[the remark this answers is not in the document]
Comment (enitrat) [reply crates/ef-testing/src/evm_sequencer/evm_state/v1.rs:217]: For now I will keep it this way, it's the same as v0, then in the future we can remove all differentiation in the same PR for both v0 v1

**Review by tcoratger — APPROVED** [review]

**Pull request merged**